### ISSUE TITLE
Memory leak when using Granian in Django

### ISSUE THREAD

**Issue by divaltor** [emmett-framework/granian#328]
Hello, thanks for your tool. It's really great, but there is problem with (possible) memory leak

Tested it with WSGI protocol on Django version 5.2, Granian 1.4.0 (and 1.4.3). In my case I upload a big video file to the server via `multipart/form-data` request and after request is done memory stay the same. I investigate it a lot and in the end found that the leak is related to Granian, because I've tried to do the same with standard `manage.py runserver` tool from Django and `gunicorn` 

Granian:
`granian --interface wsgi --host 0.0.0.0 --port 8000 server.wsgi:application`

<img width="1139" alt="image" src="https://github.com/emmett-framework/granian/assets/29837527/26ec03d0-1d4b-4034-a2d3-0365a8eef957">

Gunicorn:
`gunicorn server.wsgi`

<img width="1139" alt="image" src="https://github.com/emmett-framework/granian/assets/29837527/c4f9c3ee-9676-48ce-8ad1-f8595964b44a">

I also [attach a link](https://drive.google.com/drive/folders/1RSUJwjvYpTaJcI_T55DAF1E-MA-Fo4pS?usp=sharing) with both `.bin` files from Memray profiling for Granian and Gunicorn.

My guess that the problem is located [here](https://github.com/emmett-framework/granian/blob/80b80cbe9585789e2cf0fea977285843a14b6e5d/src/wsgi/types.rs#L125), because eventually Django read body from WSGI request and execute underline `read` method from this struct

Related Django code - https://github.com/django/django/blob/4ee68bb4f50dbe842461f8f340bb1633fddd4a53/django/core/handlers/wsgi.py#L78

**Comment by gi0baro** [comment]
Hi @divaltor, thank you for reporting this. Can you please also attach a MRE and the test method used?

**Comment by divaltor** [comment]
> Hi @divaltor, thank you for reporting this. Can you please also attach a MRE and the test method used?

@gi0baro of course. Take a look - https://github.com/divaltor/granian-mre-leak

**Comment by gi0baro** [comment]
@divaltor this should be fixed with upcoming 1.4.4

**Comment by gi0baro** [comment]
@divaltor from what I can tell from local debugging, the leak is due to the fact `wsgi.input` in environ never gets garbage collected.

I'm not sure why this happens only in Granian (at least from your investigation), but there's not so much Granian can do itself to prevent this. Also, it is not clear to me why this happens only with Django, probably some references to input are kept alive forever, and thus never GCs.
I think the best thing to do here is to open up an issue on Django repo referencing this in order to get some support from them, as of today I can't really tell why this is happening and what's the best way for Granian to prevent Django to behave like this.

**Comment by divaltor** [comment]
> @divaltor from what I can tell from local debugging, the leak is due to the fact `wsgi.input` in environ never gets garbage collected.
> 
> I'm not sure why this happens only in Granian (at least from your investigation), but there's not so much Granian can do itself to prevent this. Also, it is not clear to me why this happens only with Django, probably some references to input are kept alive forever, and thus never GCs.
> I think the best thing to do here is to open up an issue on Django repo referencing this in order to get some support from them, as of today I can't really tell why this is happening and what's the best way for Granian to prevent Django to behave like this.

First time I debugged that issue with memory leaking I found that every ref is collected by GCs in Django. May be I missed something, I'll try to debug it few more times to find why it's happening.

Appreciate a lot of your contribution on fix ❤️

**Comment by divaltor** [comment]
> ... Also, it is not clear to me why this happens only with Django, probably some references to input are kept alive forever, and thus never GCs. ...

Sorry to bother you. I've just confirmed it. Python have some delay for collecting dead references created by Django\Granian (don't know which exactly) and it so seemed to me that there was a memory leak, but after some observation and few file uploads at intervals I found that GC just need some time to remove dead references and free memory.

So there is actually no memory leak at all, thanks a lot for your awesome work ❤️
